docs/DeferredSignals.md

```diff
@@ -0,0 +1,271 @@
+# Deferred signals and why FEX needs them
+
+FEX-Emu has locations in its code which are effectively "uninterruptible". In the sense that if the guest application receives a signal during an
+"uninterruptible" code section, then FEX is likely to hang or crash in spurious and terrible ways.
+
+## Example
+FEX-Emu is in the middle of emitting code. This is a vulnerable state in which FEX can be in the middle of allocating memory or reading guest state.
+If a signal is received in the middle of this, FEX-Emu might need to compile new code in a re-entrant state. Which could involve allocating memory and
+other vulnerable things. In this case a mutex could very easily be held, interrupted, and then try to get held again while being reentrant.
+
+**This will result in a hang.**
+
+## How do we solve this?
+
+### Classical signal masking
+One solution to this problem is to mask **all** signals going in to an uninterruptible section and then unmask when leaving. This is the classical
+approach that is viable if performance isn't a significant concern. A major problem with this approach is that we must do two system calls per
+"uninterruptible" code section, which if the section is fairly small then it might cost more to ensure state integrity than doing the work at all.
+
+### Cooperative signal deferring
+A new solution is to defer asynchronous signals if they are caught inside an uninterruptible section. At the basic level, we increment a reference
+counter going in to the "uninterruptible" section, and then decrement the reference counter once we leave. This way when the signal handler receives a
+signal, it can check that thread's reference counter, store the `siginfo_t` to an array/stack object, and return to the same code segment to be handled
+later.
+
+The trick to how this works is the cost it takes to check if a signal occured, and then handling the signal safely. Ideally no signal has happened in
+the "uninterruptible" code section, so the check needs to be as cheap as possible to ensure no cost in the case that no signal has occured.
+
+The trick is that FEX maintains two memory regions for tracking deferred signals **per thread**.
+
+#### 1st memory region
+
+This region is FEX's InternalThreadState object, which is always resident for each guest thread and usually inside a register inside the JIT.
+Inside this object is where the reference counter for "uninterruptible" code segments lives. It is specifically a reference counter since these
+code segments may nest inside each other and we can only interrupt with a signal if the counter is zero.
+
+This reference counter is thread local and won't be read by any other threads, so it can be a non-atomic increment and decrement.
+Meaning it is usually three instructions (on ARM64) to increment and decrement.
+
+```cpp
+MoveableNonatomicRefCounter<uint64_t> DeferredReferenceCounter;
+MoveableNonatomicRefCounter<uint64_t> *DeferredSignalHandlerPagePtr;
```

Comment:
Isn't `DeferredSignalHandlerPagePtr` the 2nd memory region? It's misleading to place it in the `1st memory region` section then. Better to mention the variable names separately for each region.

Reply:
Resolved. Moved to the next section.